typewise-alert.cpp

```diff
@@ -11,61 +11,71 @@ BreachType inferBreach(double value, double lowerLimit, double upperLimit) {
   return NORMAL;
 }
 
-BreachType classifyTemperatureBreach(
-    CoolingType coolingType, double temperatureInC) {
-  int lowerLimit = 0;
-  int upperLimit = 0;
-  switch(coolingType) {
-    case PASSIVE_COOLING:
-      lowerLimit = 0;
-      upperLimit = 35;
-      break;
-    case HI_ACTIVE_COOLING:
-      lowerLimit = 0;
-      upperLimit = 45;
-      break;
-    case MED_ACTIVE_COOLING:
-      lowerLimit = 0;
-      upperLimit = 40;
-      break;
-  }
+BreachType classifyTemperatureBreach(CoolingType coolingType, double temperatureInC) {
+  double lowerLimit = CoolingLimitsMap[coolingType].first;
+  double upperLimit = CoolingLimitsMap[coolingType].second;
   return inferBreach(temperatureInC, lowerLimit, upperLimit);
 }
 
-void checkAndAlert(
-    AlertTarget alertTarget, BatteryCharacter batteryChar, double temperatureInC) {
+void checkAndAlert( AlertTarget *alertTarget, BatteryCharacter batteryChar, double temperatureInC) {
 
-  BreachType breachType = classifyTemperatureBreach(
-    batteryChar.coolingType, temperatureInC
-  );
+  BreachType breachType = classifyTemperatureBreach( batteryChar.coolingType, temperatureInC);
+
+  alertTarget->sendAlert(breachType);
 
-  switch(alertTarget) {
-    case TO_CONTROLLER:
-      sendToController(breachType);
-      break;
-    case TO_EMAIL:
-      sendToEmail(breachType);
-      break;
-  }
 }
 
 void sendToController(BreachType breachType) {
   const unsigned short header = 0xfeed;
-  printf("%x : %x\n", header, breachType);
+  std::cout << std::hex << header << " : " << std::hex << breachType << std::endl;
 }
 
 void sendToEmail(BreachType breachType) {
   const char* recepient = "a.b@c.com";
-  switch(breachType) {
-    case TOO_LOW:
-      printf("To: %s\n", recepient);
-      printf("Hi, the temperature is too low\n");
-      break;
-    case TOO_HIGH:
-      printf("To: %s\n", recepient);
-      printf("Hi, the temperature is too high\n");
-      break;
-    case NORMAL:
-      break;
-  }
+  std::cout << "To: " << recepient << "\n" << EmailMsgMap[breachType] << std::endl;
+}
+
+void sendToConsole(BreachType breachType){
+    std::cout << "Breach Detected: " << ConsoleMsgMap[breachType] << std::endl;
+}
+
+void EmailAlert::sendAlert(BreachType breachType)
+{
+    if (breachType != NORMAL)
+    {
+        sendToEmail(breachType);
```

Comment:
The production code (sendToEmail) and test code (m_EmailAlertSent) are together.
You have implemented a good implementation and interface separation. Why not have a separate test class for recording the m_EmailAlertSent?

Reply:
I did not quite get it. I used m_EmailAlertSent as I wanted to evaluate using the same object that was being passed to the checkAndAlert . I'd love to improve this as I realized it was not very clean but couldn't visualize other way of implementing. It would really help if you could hint with a pseudocode.(i.e if its not covered in the class today) Thanks